### DISCUSSION TITLE
XC Soar for iOS

### DISCUSSION THREAD

**Discussion by Palekf** [XCSoar/XCSoar#1351] [General]
Regarding the new EU law, which requires apple to support alternative App Stores, will XC Soar be available on iOS then (in EU)?

**Comment by MaxKellermann** [comment]
Not unless there is somebody who volunteers to maintain the iOS port. Currently, there is no volunteer.

**Comment by yorickreum** [comment]
We just brought XCSoar back into the App Store: https://apps.apple.com/us/app/xcsoar/id6741046464

**Reply by LeNinjaHD** [reply]
![image](https://github.com/user-attachments/assets/e914c670-293a-40e7-8a76-b779d028d18e)

@yorickreum I still have the bug with the waypoint icons on the latest beta, any idea what might cause this?

**Reply by yorickreum** [reply]
> @yorickreum I still have the bug with the waypoint icons on the latest beta, any idea what might cause this?

Hi @LeNinjaHD, I actually can't reproduce that ... can you provide more information about your files and settings?

![image](https://github.com/user-attachments/assets/ac896a42-c89e-46f1-866b-824e6dff968b)

**Reply by yorickreum** [reply]
Actually can you add all information on that and screenshots to this issue here? I opened it based on your earlier feedback... https://github.com/XCSoar/XCSoar/issues/1695

**Reply by LeNinjaHD** [reply]
Pretty weird. After changing some settings in System -> Looks -> Waypoints they are now normal again

I however noticed this:
![image](https://github.com/user-attachments/assets/b14f0a87-0277-44d0-b73d-836cea9f315d)

**Reply by yorickreum** [reply]
> Pretty weird. After changing some settings in System -> Looks -> Waypoints they are now normal again

Glad it fixed itself, let's keep an eye on it.

> I however noticed this:
> ![image](https://github.com/user-attachments/assets/b14f0a87-0277-44d0-b73d-836cea9f315d)

What do you mean? The weird icons around the scale bar?

**Reply by yorickreum** [reply]
Actually, these smiley faces seem to appear at specific zoom levels. I wonder if that is intended, if too many obstacle icons would be rendered very close to each other they just turn into a smiley?

**Reply by kobedegeest** [reply]
Those smileys are an icon used for your "friends" you track via skylines. So I think it picks the wrong icon 
on android at large zoom icons do become a dot when zoomed in you get the specific type of waypoint. But never a smiley face

**Reply by TheMiNuS** [reply]
Thnaks a lot for making this app available on Apple devices.
I tried it in flight and it worked well, with the heavy power consumption that you already shared as something to improve.

The thing that could be great to be fixed in a near version is the fact that when you lock the phone to save power or to put it on the side to do something else, when you reactivate the screen the app just restart from scratch and ask for fly/nav instead of recovering it's previous state while running in background.

Is it something already identified or planned to be fixed ?

**Comment by lordfolken** [comment]
Congratulations on making it into the store. :+1

A few questions:
 * There are currently 3 PRs open for IOS https://github.com/XCSoar/XCSoar/labels/IOS
The binary that is published, is  that the current head, or is that a local branch of yours? (please be careful GPL requires to make the exact source availble.) 
 * Would you like to officialize this app? (i.e. link it on the webpage, readme, official platform etc.?)

**Reply by yorickreum** [reply]
> Congratulations on making it into the store. :+1

Thanks! (Also for the continued support on the way here on GitHub...)

> There are currently 3 PRs open for IOS 
> The binary that is published, is  that the current head, or is that a local branch of yours? (please be careful GPL requires to make the exact source availble.)

The published binary is built from the current head of the official repository. That is why I was pushing a bit recently to get the mission-critical PRs merged. The PRs still open can follow in an update. I would like to try to keep the app store binary based on and up-to-date with the official repository (and not maintain my own fork).

> Would you like to officialize this app? (i.e. link it on the webpage, readme, official platform etc.?)

Sounds good to me! I thoroughly hope we can keep maintaining (and improving) the iOS version from now on.

**Reply by XCNav** [reply]
Just shared it on FB in our XCSoar Group with more than 5000 users

https://www.facebook.com/share/p/19dXyfg47N/

Most of the basic functions are working with our XCRemote connected via BLE without problems , I will share a Video tonight in the FB group. However, Left/Right arrow seems not to work in IOS for changing through the pages. Is there any other key command for that?
As well sending NMEA commands via XCI seems not to be working.
Rgds Markus
XCNAV

**Comment by WolfgangDrescher** [comment]
Is it documented somewhere how to set up a development environment on macOS to help with the open issues? I couldn't find anything on how to create the iOS build and test it (on macOS). Since I am new to the XCSoar codebase, but would be interested in helping with the iOS migration, I would appreciate a good introduction.

**Reply by WolfgangDrescher** [reply]
I have now made progress and managed to build with the command I found in #1625, but the build gets stuck at `dither`.

```
gmake -j$(nproc) TARGET=IOS64  V=2 OPTIMIZE="-O0" ipa
dither -type GrayScale -define png:color-type=0 Data/bitmaps/aboveterrain.bmp output/data/bitmaps/aboveterrain.png
gmake: dither: No such file or directory
gmake: *** [build/resource.mk:24: output/data/bitmaps/aboveterrain.png] Error 127
gmake: *** Waiting for unfinished jobs....
```

**Reply by yorickreum** [reply]
Thanks for supporting the iOS build, @WolfgangDrescher!

I had added some documentation on my development environment in https://github.com/XCSoar/XCSoar/pull/1713
It's not merged yet.

You can also check the log of the GitHub Actions CI; they build successfully for iOS and macOS.

I have not seen the dither error before ... seems odd. First thing I'd try is just running `gmake clean` and try again... If it still doesn't work I'd need some more information on the environment and the build log to investigate.

**Reply by yorickreum** [reply]
Is ImageMagick installed and in the path? Which version? In the line of resource.mk that throws the error actually the ImageMagick `convert` command is supposed to be called, dither is just an argument there.

**Reply by WolfgangDrescher** [reply]
Thank you, this helped. I had to `brew install imagemagick` and `brew install sox` was also needed.

Should we add `sox` and `imagemagick` to `install-darwin-packages.sh`?

`make TARGET=IOS64 ipa` actually throws an error:

```
In file included from ./src/Geo/GeoPoint.hpp:7:
./src/Math/Classify.hpp:24:10: error: use of infinity is undefined behavior due to the currently enabled floating-point options [-Werror,-Wnan-infinity-disabled]
   24 |   return __builtin_isfinite(value);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
./src/Geo/GeoPoint.hpp:82:12: note: in instantiation of function template specialization 'IsFinite<double>' requested here
   82 |     return IsFinite(longitude.Native()) &&
      |            ^
1 error generated.
make: *** [output/IOS64/dbg/src/Airspace/ProtectedAirspaceWarningManager.o] Error 1
```

However, when using `gmake`, this issue doesn’t appear. Maybe it's worth mentioning `gmake` in your updated in `doc/build.rst`.

---

At one point I got:

```
A required plugin failed to load. Please ensure system content is up-to-date — try running 'xcodebuild -runFirstLaunch'.
```

So it seems that XCode needs to be installed for the build to work. I'll keep trying things out and see if I can get it to work. I'll come back to you once XCode is installed and I still could not get it running. Thanks @yorickreum for the fast support!

---

Is #1705 already published in the current version in the AppStore? #1693 seems to be the biggest issue. My battery is drained by 1% every minute and the phone gets so hot that I'm afraid it won't survive a hot day in the cockpit.

**Reply by WolfgangDrescher** [reply]
The `IOS64` and `IOS64SIM` builds seem to complete now, but I cannot get them running. Am I looking at the correct directory in `output/IOS64` / `output/IOS64SIM`? Should I use the `xcsoar.ipa` file or `output/IOS64SIM/ipa/Payload/XCSoar.app`? The IPA file warns me with `Diese App kann nicht installiert werden, da ihre Integrität nicht verifiziert werden konnte`. I guess the app signing is missing? The XCSoar.app file won’t start because `XCSoar wurde unerwartet beendet`. Probably it is me, because I am not an iOS developer. Do you have any idea how I could debug further?
Building with your `XCSoar.xcodeproj` from #1713 did not work either, but I did not investigate this too much so far.

**Reply by yorickreum** [reply]
> Should we add `sox` and `imagemagick` to `install-darwin-packages.sh`?

Sounds like a good idea to me.
 
> However, when using `gmake`, this issue doesn’t appear. Maybe it's worth mentioning `gmake` in your updated in `doc/build.rst`.

Yeah, on my system make and gmake actually resolve to GNU make (just different versions). What does make resolve to for you? `which make` and `make -v` would be interesting.

> So it seems that XCode needs to be installed for the build to work. I'll keep trying things out and see if I can get it to work. I'll come back to you once XCode is installed and I still could not get it running. Thanks @yorickreum for the fast support!

Might be, iOS development pretty much always requires at least the XCode command line tools.
 
> Is #1705 already published in the current version in the AppStore? #1693 seems to be the biggest issue. My battery is drained by 1% every minute and the phone gets so hot that I'm afraid it won't survive a hot day in the cockpit.

Yes, #1705 is merged in the currently published version. Sadly it only helped a tiny bit if at all with #1693 (but also solved some other issues).

>  Should I use the xcsoar.ipa.

Yes. This is the same as in the ipa directory, but packed (zipped) so it can be installed. 

> Probably it is me, because I am not an iOS developer. Do you have any idea how I could debug further?

Yes. This is to be expected, as Apple requires any executable to be signed (with a certificate from an Apple developer account), unsigned executables can only be installed on jailbroken devices – or the simulator. Hence I'd recommend to test on the iOS simulator with the .ipa from IOS64SIM. 
If you want to test it on your own device and not get your own Apple developer account (it's $ 100 / year), I can maybe provide you with a team signing certificate for that purpose.

**Reply by WolfgangDrescher** [reply]
<details>
<summary>Click to see my `make` and `gmake` setup</summary>

### `which make`

```
/usr/bin/make
```

### `which gmake`

```
/opt/homebrew/bin/gmake
```

### `make -v`

```
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

### `gmake -v`

```
GNU Make 4.4.1
Built for aarch64-apple-darwin24.0.0
Copyright (C) 1988-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

</details>

---

I'm slowly getting there, thanks again for your help. I can open XCSoar now in the Simulator. I could drag & drop the unsigned IOS64SIM XCSoar.app into the simulator. First I thought that everything was bugged, but then I realized that I have to set a custom location for the GPS in the simulator.

I'll come back to you for testing on my real device once I feel more comfortable with the build and debug workflow.

The bug with the wrong icons is somewhat reproducible in the simulator. Smileys won't appear, but the icons for waypoints are broken. So I'll try to take this as a starting point to learn the XCSoar codebase as well as how to debug in the simulator.

It would be nice to add a BluetoothHelper for Apple devices. Should I open a new issue for this?

**Reply by yorickreum** [reply]
> Should we add sox and imagemagick to install-darwin-packages.sh?

I just checked, it's already there, in the `BASE` section.

> It would be nice to add a BluetoothHelper for Apple devices. Should I open a new issue for this?

That would be great.

**Reply by BrunoJ-WN** [reply]
Hello,
I use XCSOAR on an Android tablet in my glider, and I would also like to run it on my iPhone. Since Apple restricts access to the iPhone's internal files, is there a way to copy the profile files and waypoint databases directly from an SD card or computer, as you can on an Android device ?
Thanks a lot in advance.

**Reply by yorickreum** [reply]
> Hello,
> 
> I use XCSOAR on an Android tablet in my glider, and I would also like to run it on my iPhone. Since Apple restricts access to the iPhone's internal files, is there a way to copy the profile files and waypoint databases directly from an SD card or computer, as you can on an Android device ?
> 
> Thanks a lot in advance.

You can just use the _Files_ app on your iPhone, click on "On my iPhone" -> "XCSoar".

**Reply by BrunoJ-WN** [reply]
Thank you very much for your quick response. It is indeed very easy and obvious. I am usually smarter than that, but this time I was wrong! I was misled by the fact that it displays a line /private/var/mobile/Containers/Data/Application………typical of hidden files.
Everything is perfect.
Bruno

**Reply by yorickreum** [reply]
> Thank you very much for your quick response. It is indeed very easy and obvious. I am usually smarter than that, but this time I was wrong! I was misled by the fact that it displays a line /private/var/mobile/Containers/Data/Application………typical of hidden files.

No worries. It is indeed confusing and useless to show the raw Unix path on iOS. We could improve that.

**Comment by erock999** [comment]
Is it a known issue ( or some other settings required? ) that Replay function on 7.44 seemed time still/hold and updates when touch the screen.

**Reply by erock999** [reply]
Fixed in 7.44 released on 23rd March 2026.
